app/src/main/java/com/palone/planahead/domain/alertEngine/AlertEngineImpl.kt

```diff
@@ -0,0 +1,40 @@
+package com.palone.planahead.domain.alertEngine
+
+import com.palone.planahead.data.database.AlertRepository
+import com.palone.planahead.data.database.TaskRepository
+import com.palone.planahead.data.database.alert.Alert
+import com.palone.planahead.data.database.task.Task
+import com.palone.planahead.data.database.task.properties.TaskSource
+import com.palone.planahead.data.database.task.properties.TaskType
+import com.palone.planahead.services.alarms.AlarmsHandler
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.last
+import kotlinx.coroutines.launch
+
+class AlertEngineImpl(
+    private val taskRepository: TaskRepository,
+    private val alarmsHandler: AlarmsHandler,
+    private val alertRepository: AlertRepository
+) : AlertEngine {
+    override val scope = CoroutineScope(Dispatchers.IO)
+    override fun createDelayedOneTimeAlert(
+        alert: Alert,
+        task: Task,
+        delay: Long
+    ) {
+        scope.launch {
+            val currentEventMillisInEpoch = alert.eventMillisInEpoch ?: 0
+            val newTask =
+                task.copy(taskId = null, taskType = TaskType.ONE_TIME, source = TaskSource.SYSTEM)
+            val newTaskId = taskRepository.upsert(newTask)
+            val newAlert = alert.copy(
+                alertId = null,
+                taskId = newTaskId.toInt(),
+                eventMillisInEpoch = currentEventMillisInEpoch + delay
+            )
+            alertRepository.upsert(newAlert)
```

Comment:
It would be better if `upsert` just returned the newly created instance. In the future we can have a sync process that can add tasks/alerts and last one will not be guaranteed to be the one we just saved.